src/game/items.c

```diff
@@ -416,28 +388,95 @@ void Item_AlignPosition(
 }
 
 bool Item_MovePosition(
-    ITEM_INFO *src_item, ITEM_INFO *dst_item, PHD_VECTOR *vec, int32_t velocity)
+    ITEM_INFO *item, const ITEM_INFO *ref_item, const PHD_VECTOR *vec,
+    int32_t velocity)
 {
-    PHD_3DPOS dst_pos;
-    dst_pos.x_rot = dst_item->pos.x_rot;
-    dst_pos.y_rot = dst_item->pos.y_rot;
-    dst_pos.z_rot = dst_item->pos.z_rot;
+    const PHD_3DPOS *ref_pos = &ref_item->pos;
     Matrix_PushUnit();
-    Matrix_RotYXZ(
-        dst_item->pos.y_rot, dst_item->pos.x_rot, dst_item->pos.z_rot);
+    Matrix_RotYXZ(ref_pos->y_rot, ref_pos->x_rot, ref_pos->z_rot);
     MATRIX *mptr = g_MatrixPtr;
-    dst_pos.x = dst_item->pos.x
-        + ((mptr->_00 * vec->x + mptr->_01 * vec->y + mptr->_02 * vec->z)
-           >> W2V_SHIFT);
-    dst_pos.y = dst_item->pos.y
-        + ((mptr->_10 * vec->x + mptr->_11 * vec->y + mptr->_12 * vec->z)
-           >> W2V_SHIFT);
-    dst_pos.z = dst_item->pos.z
-        + ((mptr->_20 * vec->x + mptr->_21 * vec->y + mptr->_22 * vec->z)
-           >> W2V_SHIFT);
+
+    PHD_3DPOS dst_pos = {
+        .x = ref_pos->x
+            + ((mptr->_00 * vec->x + mptr->_01 * vec->y + mptr->_02 * vec->z)
+               >> W2V_SHIFT),
+        .y = ref_pos->y
+            + ((mptr->_10 * vec->x + mptr->_11 * vec->y + mptr->_12 * vec->z)
+               >> W2V_SHIFT),
+        .z = ref_pos->z
+            + ((mptr->_20 * vec->x + mptr->_21 * vec->y + mptr->_22 * vec->z)
+               >> W2V_SHIFT),
+        .x_rot = ref_pos->x_rot,
+        .y_rot = ref_pos->y_rot,
+        .z_rot = ref_pos->z_rot,
+    };
+
     Matrix_Pop();
 
-    return Item_Move3DPosTo3DPos(&src_item->pos, &dst_pos, velocity, MOVE_ANG);
+    {
+        const int32_t dx = dst_pos.x - item->pos.x;
+        const int32_t dy = dst_pos.y - item->pos.y;
+        const int32_t dz = dst_pos.z - item->pos.z;
+        const int32_t dist = Math_Sqrt(SQUARE(dx) + SQUARE(dy) + SQUARE(dz));
+        if (velocity >= dist) {
+            item->pos.x = dst_pos.x;
+            item->pos.y = dst_pos.y;
+            item->pos.z = dst_pos.z;
+        } else {
+            item->pos.x += (dx * velocity) / dist;
+            item->pos.y += (dy * velocity) / dist;
+            item->pos.z += (dz * velocity) / dist;
+        }
+    }
```

Comment:
Why the floating `{` `}` around this chunk of code? Because `dx` and `dz` are reused later?

Reply:
Yes. Is it awful?

Reply:
It could possibly be confusing, but idk better names off the top of my head lol.